client/components/tinymce/plugins/media/plugin.jsx

```diff
@@ -234,32 +348,45 @@ function mediaButton( editor ) {
 				return;
 			}
 
-			// The `img` object can be a string or a DOM node. We need a
-			// normalized string to replace the post content markup
-			let imgString;
-			if ( img.outerHTML ) {
-				imgString = img.outerHTML;
+			let mediaString = '';
+
+			if ( mediaHasCaption ) {
+				if ( captionNode.outerHTML ) {
+					mediaString = captionNode.outerHTML;
+
+					captionNode.outerHTML = event.content;
```

Comment:
This is how we update the post/page editor image.

Now, if the updated image also has a caption, let's target the caption wrapper and replace its `outerHTML` with the new markup. Otherwise, continue as before.